docs/architecture/adr-038-state-listening.md

```diff
@@ -204,60 +204,158 @@ func (rs *Store) CacheMultiStore() types.CacheMultiStore {
 requests that affected them and the ABCI responses they affected. The `ABCIListener` interface also exposes a
-`ListenSuccess` method which is (optionally) used by the `BaseApp` to await positive acknowledgement of message
-receipt from the `StreamingService`.
+`ListenKVStorePair` method which is used by the `StoreKVPairWriteListener` to stream state change events.
 
 ```go
 // ABCIListener interface used to hook into the ABCI message processing of the BaseApp
 type ABCIListener interface {
-    // ListenBeginBlock updates the streaming service with the latest BeginBlock messages
-    ListenBeginBlock(ctx types.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error
-    // ListenEndBlock updates the steaming service with the latest EndBlock messages
-    ListenEndBlock(ctx types.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error
-    // ListenDeliverTx updates the steaming service with the latest DeliverTx messages
-    ListenDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
-    // HaltAppOnDeliveryError whether or not to halt the application when delivery of massages fails
-    // in ListenBeginBlock, ListenEndBlock, ListenDeliverTx. When `false, the app will operate in fire-and-forget mode.
-    // When `true`, the app will gracefully halt and stop the running node. Uncommitted blocks will
-    // be replayed to all listeners when the node restarts and all successful listeners that received data
-    // prior to the halt will receive duplicate data. Whether or not a listener operates in a fire-and-forget mode
-    // is determined by the listener's configuration property `halt_app_on_delivery_error = true|false`.
-    HaltAppOnDeliveryError() bool
-}
-
-// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
-type StreamingService interface {
-    // Stream is the streaming service loop, awaits kv pairs and writes them to a destination stream or file
-    Stream(wg *sync.WaitGroup) error
-    // Listeners returns the streaming service's listeners for the BaseApp to register
-    Listeners() map[types.StoreKey][]store.WriteListener
-    // ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
-    ABCIListener
-    // Closer interface
-    io.Closer
+	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
+	ListenBeginBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenEndBlock updates the steaming service with the latest EndBlock messages
+	ListenEndBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
+	ListenDeliverTx(blockHeight int64, req []byte, res []byte) error
+	// ListenStoreKVPair updates the steaming service with the latest StoreKVPair messages
+	ListenStoreKVPair(blockHeight int64, data []byte) error
```

Comment:
```diff
 type ABCIListener interface {
- 	ListenBeginBlock(ctx types.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error
+	ListenBeginBlock(blockHeight int64, req []byte, res []byte) error
-	ListenEndBlock(ctx types.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error
+	ListenEndBlock(blockHeight int64, req []byte, res []byte) error
-	ListenDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
+	ListenDeliverTx(blockHeight int64, req []byte, res []byte) error
 	ListenStoreKVPair(blockHeight int64, data []byte) error
 }
```

You've changed this interface so that its methods no longer use well-defined types, and now use raw bytes. This is pretty strange: it somewhat subverts the type system, it adds a dimension of complexity and risk that wasn't there before, and it moves a lot of responsibility to the implementation. Why is it necessary?

Reply:
> You've changed this interface so that its methods no longer use well-defined types, and now use raw bytes. This is pretty strange: it somewhat subverts the type system, it adds a dimension of complexity and risk that wasn't there before, and it moves a lot of responsibility to the implementation. Why is it necessary?

What I am going for is a simple dumb pipe that streams the data out. I want an interface that avoids extra dependencies in the plugin implementation. I want an interface that is consistent across SDK versions. I want a pipe that is consistent and is not effected by a proto format change that occurs during a block upgrade. I want separation of concern.

> _edit_: maybe [#13473 (comment)](https://github.com/cosmos/cosmos-sdk/pull/13473#discussion_r996609406) ?

Changing the interface is what made this possible but not the reason.

Reply:
> What I am going for is a simple dumb pipe that streams the data out. I want an interface that avoids extra dependencies in the plugin implementation. 

The interface isn't defined in terms of dumb bytes, it's defined in terms of well-defined domain concepts such as BeginBlock, EndBlock, etc. (Also, streaming would require io.Reader and io.Writer, not []byte.) 

> I want a pipe that is consistent and is not effected by a proto format change that occurs during a block upgrade. I want separation of concern.

If I write an implementation of the proposed interface that works with proto version X, and the proto version changes to X+1, then my implementation will continue to satisfy the interface, but it will no longer be compatible with callers who send the new version. This will compile, but fail at runtime, which will likely crash my node. Is this a problem?

If implementations are expected to decode _all possible_ encodings of the corresponding type, then this needs to be well documented, tested, etc. etc. because it is highly nonidiomatic.

Reply:
After catching up with @egaxhaj on today's SDK call there is a sensible use case for transmitting raw bytes to a listener.  This use case writes bytes without unmarshaling and a downstream process is responsible for keeping up with the proto definitions required to unmarshal instead of placing the maintenance burden on the plugin.

However I do agree this non-idiomatic in the current SDK, and `[] byte` in the interface definition for **all** streaming implementations seems like not a good assumption.  Could we move the marshaling step into the listener implementation which satisfies your plugins contract / use case and keep the interface as originally specified?

Reply:
> There is a sensible use case for transmitting raw bytes to a listener. This use case writes bytes without unmarshaling and a downstream process is responsible for keeping up with the proto definitions required to unmarshal instead of placing the maintenance burden on the plugin.

That's fine, but it means that the SDK defers all responsibilities related to schema management to the listener. In that case, there's no reason for the ABCIListener interface on the SDK side to define discrete methods per message type. Those types are part of the schema. The interface should instead be something like

```go
// If type information is part of the encoded values.
type ABCIListener interface {
    Listen(ctx context.Context, req, res []byte) error
}

// If type information is not part of the encoded values.
type ABCIListener interface {
    Listen(ctx context.Context, type string, req, res []byte) error
}
```

Reply:
> there's no reason for the ABCIListener interface on the SDK side to define discrete methods per message type

Why not keep `ABCIListener` implemented with schema methods but delegate the task of satisfying the schema-free plugin contract (that is `[]bytes`) to the associated listener?

Reply:
> Why not keep ABCIListener implemented with schema methods but delegate the task of satisfying the schema-free plugin contract (that is []bytes) to the associated listener? I'm picturing an architecture that allows for other plugins to rely on the proto schema.

Because — AFAIU, correct me if I'm wrong — the proto schema includes both methods and types. If you add a method, it's a new (version of) the schema, which old implementations don't satisfy. So defining an interface called ABCIListener isn't sufficient, right? It's got to be ABCIListenerV1, ABCIListenerV2, etc. where the V's are updated whenever there is any meaningful change to the proto schema.

Reply:
> So if you want to have a specific set of methods as part of an interface that's just called ABCIListener without any version, that isn't sufficient, right?

yeah, we need some handshake api then, to reject plugins early on if it don't support the version.

Reply:
> yeah, we need some handshake process then, to reject the plugins early on if it don't support the version.

What does rejection imply? Does it crash the node? Or will the node retry the connection until it succeeds? What should happen if a plugin service claims it supports version X during handshaking but rejects those messages during runtime?

> We should distinguish the intra-process and inter-process interfaces, we should have typed interface for intra-process method calls, but untyped ones for inter-process communication, for example the inter-process grpc service interface could be an untyped one, adding the format negotiation, could be something like this:

Doesn't gRPC itself already do the format (i.e. schema) negotiation you're approximating with this service definition?

Reply:
> What does rejection imply? Does it crash the node? Or will the node retry the connection until it succeeds?

the safest way is probably refuse to startup.

> What should happen if a plugin service claims it supports version X during handshaking but rejects those messages during runtime?

the error result should be handled like any other errors, either stop the consensus state machine, or ignore it.

Reply:
> Doesn't gRPC itself already do the format (i.e. schema) negotiation you're approximating with this service definition?

this interface can support the "dumb pipe" that just claims supporting all formats and pass the payloads to downstream.

Reply:
> the safest way is probably refuse to startup.

Say the other process handshakes OK when the node starts up, but 2 minutes later is restarted with a different binary that handshakes incorrectly. Should the node itself then crash? How does the node distinguish between a bad other process and a faulty connection to that process?

> the error result should be handled like any other errors, either stop the consensus state machine, or log and ignore.

These are two vastly different outcomes!

Reply:
> this interface can support the "dumb pipe" that just claims supporting all formats and pass the payloads to downstream.

In this case, you're deferring all error checking to runtime, and the interface reduces to an io.Writer.

Reply:
> > this interface can support the "dumb pipe" that just claims supporting all formats and pass the payloads to downstream.
> 
> In this case, you're deferring all error checking to runtime, and the interface reduces to an io.Writer.

for inter-process communications, it's always checked at runtime, the difference is if we rely on low level software like grpc service to do that, or reinvent some wheels, depending on our requirements.

Reply:
> > the safest way is probably refuse to startup.
> 
> Say the other process handshakes OK when the node starts up, but 2 minutes later is restarted with a different binary that handshakes incorrectly. How would it detect that change? Should the node itself then crash? Can the node distinguish between a bad other process and a faulty connection to that process?

I see, you mean the plugin is restart and reconnect at runtime, it seems this part is missed from current ADR, I would suggest something like this:
- the plugin processes are always loaded and started when node startup
- if a plugin quit at runtime, it's either ignored or stop the node together, depending on the option `StopNodeOnErr`.
- one can not start a plugin service at runtime and connect to the node.

> 
> > the error result should be handled like any other errors, either stop the consensus state machine, or log and ignore.
> 
> These are two vastly different outcomes!

yeah, that's user's choice to make, depending on the use cases, the `StopNodeOnErr` option in this ADR is for this.

Reply:
I've pointed out just a tiny fraction of the issues that can occur, and can go on basically forever ;) That's because this is a formally hard problem, which you're not gonna solve with a novel design in a GitHub issue, and then addressing individual issues one by one. That's OK! You shouldn't need to; the SDK should definitely not try to do any of this stuff from first principles. There are already reasonable solutions you can use. If you want plugins, you should use HashiCorp's go-plugin. They basically know what they're doing.

Reply:
>edit: It's also worth pointing out that Listen will return an error if the sender ever sends bytes which the listener doesn't know how to parse. That will happen any time there's a new schema definition, and it's deployed to the sender (the Cosmos node) before it's deployed to the receiver (the listener). Which means you have to ensure, somehow, that changes are always deployed to the listeners before they're deployed to the nodes.

Yes, it is definitely a hard problem which I don't think a solution exists just different approaches. I'll try and document this as best I can in the ADR and would appreciate any help with the wording below.

>Our decision to break away from using the typed system in the ABCIListener is for the following reason:
>
>If we stick with defined types in the ABCIListener interface methods we force plugin implementations to keep up with update on every proto schema change. This forces teams to check every SDK update for changes to the proto schema. Not fun at all and definitely avoidable.
>
>Our solution is to use a typeless interface that streams the raw []byte data to plugins.
This allows plugin implementations to operate without worry but only if they avoid any proto parsing in the plugin implementation (this is key). Therefore, the plugins role becomes that of solely ingesting the raw data to an external system (DB, RebitMQ, Kafka, etc) and make it the responsibility of downstream systems to handle multi version proto schema parsing (blockchain explorers already do this). This approach provides the following benefits:
> - no more having to check for changes on every version update
> - no more having to rush to update a plugin on proto schema update
> - because plugins avoid parsing of data, nodes can continue to operate efficiently without worry of plugins hindering overall preference, which may lead to nodes being dropped from the network

cc: @peterbourgon

Reply:
If the ABCIListener will be a "dumb pipe" to other systems, that's cool. But in that case it should be a true dumb pipe, with no knowledge of the data flowing through it. There's interfaces for that already: io.Reader and io.Writer. 

It could alternatively be semantically aware, and implement specific ABCI methods with corresponding types. This interface would necessarily be versioned, and that versioning by definition includes not only the types but also the methods of the interface.

What it shouldn't be is a hybrid of these two approaches: an interface with a specific set of named methods, but without any typed parameters either accepted or returned by those methods.

If the goal is to allow the req/res parameters to change without breaking implementations, gRPC already solved this for you natively: something that takes a FooRequestV1 will be able to take a FooRequestV2 if you follow the rules. And you are obliged to follow the rules, because if you don't you break everyone.

If the goal is something else, please clarify!

Reply:
I propose `ABCIListener` maintain types.

```golang
type ABCIListener interface {
	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
	ListenBeginBlock(ctx types.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error
	// ListenEndBlock updates the steaming service with the latest EndBlock messages
	ListenEndBlock(ctx types.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error
	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
	ListenDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
}
``` 

Folks can implement this interface as dumb pipe if they want to.

```golang
var (
	_ ABCIListener = (*DumbPipeABCIListener)(nil)
)

type DumbPipeABCIListener struct {
	writer io.Writer
}

func (d DumbPipeABCIListener) ListenBeginBlock(ctx types.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error {
	reqbz, _ := req.Marshal()
	resbz, _ := req.Marshal()
	d.writer.Write(reqbz)
	d.writer.Write(resbz)
}
```

A corresponding plugin can operate with a listener if desired (in this case a dumb pipe plugin).  I don't think the 2 entities (listeners and plugins) ought to be tightly coupled.  

This also allows other users (implementers) to opt for strong proto buff typing throughout the pipeline, which is a good idea in my opinion, but that's also their choice.

Reply:
@egaxhaj 

> If we stick with defined types in the ABCIListener interface methods we force plugin implementations to keep up with update on every proto schema change. This forces teams to check every SDK update for changes to the proto schema. Not fun at all and definitely avoidable.

I don't think this is accurate. Protobuf schemas are expected to be backwards compatible. If a given protobuf schema changes, plugins using a prior version should continue to work fine, modulo whatever new information in the updated schema they might be missing. If a change isn't backwards compatible, then either (a) it needs a new identifier, or (b) maintainers must ensure that the changes won't break anything in prod, so to speak, which Cosmos definitely can't do.

Reply:
😄 I think we're at a concuss. No I never liked the idea of coupling both a dumb pipe and well defined types. The two interface approach is excellent.

We continue to maintain the `ABCIListener` interface as is. 

```
type ABCIListener interface {
	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
	ListenBeginBlock(ctx types.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error
	// ListenEndBlock updates the steaming service with the latest EndBlock messages
	ListenEndBlock(ctx types.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error
	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
	ListenDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
}
```

And then in the plugin if you want a dumb pipe you can do something like

```
type ABCIListenerPlugin interface {
	Listen(ctx context.Context, blockHeight int64, eventType string, data []byte) error
}
```

And the implementation

```
var (
	_ baseapp.ABCIListener  = (*GRPCClient)(nil)
	_ ABCIListenerPlugin    = (*GRPCClient)(nil)
)

// GRPCClient is an implementation of the ABCIListener and ABCIListenerPlugin interfaces that talks over RPC.
type GRPCClient struct {
	client ABCIListenerServiceClient
}
```

@yihuang I don't know how to pass block info to the `io.Writer` in your `DumbPipeABCIListener` example. Either way this is up to the plugin authors now.

@yihuang @peterbourgon Take a look at the draft implementation https://github.com/cosmos/cosmos-sdk/pull/13472/commits/fcd438ba3089fc756f1290b52c97d63eb03a2111

@peterbourgon I also address the issue in https://github.com/cosmos/cosmos-sdk/pull/13473#discussion_r998568668

The `WriteListener` becomes:

```
type WriteListener interface {
	OnWrite(storeKey StoreKey, key []byte, value []byte, delete bool)
}
```

and introduced a new `StoreListener` interface in the `baseapp/streaming.go`.  The same concept as the `ABCIListener` interface. Although, this could be rolled up into the ABCIListener interface. We can go either way here but let me know your thoughts on this (If they're separate you could decide to listen to just store state changes).

https://github.com/cosmos/cosmos-sdk/blob/fcd438ba3089fc756f1290b52c97d63eb03a2111/baseapp/streaming.go#L14-L85

GRPC service for the Dumb pipe approach.

https://github.com/cosmos/cosmos-sdk/blob/fcd438ba3089fc756f1290b52c97d63eb03a2111/streaming/plugins/abci/proto/v1/abci_listener.proto#L12-L22



One more questions here. Do `tx`s and state changes arrive in a deterministic order? If either are true, we will need to preserver order. Of course, this only applies when consistency is desired. In that case we'll need to add a `Stream` method to the ABCIListener interface that can be called when `stop-node-on-err=true`:

```go
type ABCIListener interface {

    ...

    StreamDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
}
```

Reply:
I am not seeing this interface as an improvement over `ABCIListener` it's the same parameters arranged slightly differently. 

```golang
type ABCIListenerPlugin interface {
	Listen(ctx context.Context, blockHeight int64, eventType string, data []byte) error
}
```

 A plugin specific implementation ought to be sufficient.  Am I missing something?

Reply:
We're in agreement that `ABCIListener` interface is staying as is. `ABCIListenerPlugin` interface is **plugin implementation specific**. It servers as a bridge on how to build a `dump pipe` that helps avoiding [protocol breaking changes](https://learn.microsoft.com/en-us/aspnet/core/grpc/versioning?view=aspnetcore-6.0#behavior-breaking-changes). I'll be doing a new write up of the ADR soon and will only focus on the `ABCIListener` implementation as example.

Reply:
ref: https://github.com/cosmos/cosmos-sdk/pull/13516#issuecomment-1293587418

I'm in agreement with @yihuang. If we remove the `WriteListener` interface and introduce an internal `MemoryListener`. Than we can do something along the lines of...

```
func (app *BaseApp) Commit() (res abci.ResponseCommit) {
    ...
    
    if app.abciListener != nil {
        cache := app.cms.PopStateChange()
        app.abciListener.ListenCommit(ctx, res, cache)
    }
}
```


Interface
```
type ABCIListener interface {

    ...

    ListenCommit(ctx types.Context, res abci.ResponseCommit, changeSet []store.StoreKVPair) error
}